### PR TITLE
ABR4: repairInternalTables

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**: Start copying restore tasks over to AtlasDB

**Implementation Description (bullets)**:
- Added AtlasRestoreTask.repairInternalTables

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added ETE test to verify that the ranges we get are of the correct form (even in 3-node RF2 setup).
There are more lower-level tests (unit tests) to be copied over from IBP (the internal backup product), but I'm deferring that to another PR due to the magnitude of this change

+1228 lines, consisting of:
- 188 lines of class-uniqueness
- 131 lines of refactor  (ClusterFactory, DefaultCqlClientFactory)
- a few hundred lines copied directly or paraphrased from IBP (CassandraRepairHelper, ClusterMetadataUtils, CqlCluster)
- some wiring (AtlasRestoreService + misc classes with small changes)
- a few test classes

**Concerns (what feedback would you like?)**:
Is the public API OK? Does the consumer look right?

**Where should we start reviewing?**:
CassandraRepairHelper

**Priority (whenever / two weeks / yesterday)**: today, hopefully

### CODE REVIEW

File: atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/cassandra/ClusterMetadataUtils.java
Comment (gsheasby) [inline atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/cassandra/ClusterMetadataUtils.java:42]: This was copied over from the internal backup product - I haven't yet made an attempt to refactor the code. There are tests internally that we should also copy over.

File: atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/cassandra/CqlCluster.java
Comment (gsheasby) [inline atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/cassandra/CqlCluster.java:127]: utils copied over from internal backup product, but hopefully will not exist in that product once we're done.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:78]: we probably need to check this with a sample of production config if we wanted to assume CQL support. I'd be a bit concerned because we don't mandate this in the general case, but _maybe_ most people do have it

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:40]: up to you, though at some point here I'd probably add some marker interfaces

**Review by jeremyk-91 — COMMENTED** [review]
I think we do need to discuss the general choice here of using CQL where Thrift was previously used and what we want to do about that longer-term. I haven't gone into full detail yet - maybe let's discuss this next week when everyone's around?

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java
Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java:25]: I moved `SocksProxyNettyOptions` outside this package so that I could use it elsewhere.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:116]: yeah, I didn't get bandwidth last week to refactor the copied code.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:120]: this part was also copied from internal source - can definitely #be-better here and calculate numTokenRanges only if needed.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:40]: yep, likely to do that kind of thing once the structure of this code becomes apparent - which will probably not be until the further tasks are added.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:78]: so the internal code assumes CQL support (it creates a Cluster object and then does CQL queries (I copied this logic over to CqlCluster).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/client/creation/ClusterFactory.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/client/creation/ClusterFactory.java:41]: Pure refactor - this was extracted from DefaultCqlClusterFactory (see below).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:80]: So the tokens returns aren't the same - they're a proper subset of the token map, except in the edge case where the end point of each range in the token ring has been hit.

I've added a test that shows that the ranges returned are of the form that we expect.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:125]: This method can do with some nice refactoring. I would rename it to `getMinimalSetOfRangesForTokens`

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:144]: extract a method `findTokenRange`

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:153]:
```suggestion
            tokenRangesByStartToken.merge(
                    smallTokenRange.getStart(), smallTokenRange, ClusterMetadataUtils::findLatestEndingRange);
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:80]: It is probably worth it to use this code, as we do need to migrate to CQL eventually anyway

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:117]: This will add the token range to the map for each host, so the javadoc is lying? (note that the internal product seems to do the same thing). That said, maybe this is what we want?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:119]: 🎉

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:120]: I prefer map to stream then map to list

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:80]: Overall this class is a faithful representation of the internal product, but it can definitely benefit from a moderate refactor as outlined below. Also, please add unit tests!

**Review by gmaretic — COMMENTED** [review]
Publishing this partial review to unblock. Will continue after lunch

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:99]: Hmm, I think I prefer extracting a helper method instead of the long sequence of mapKeys:
```
        return KeyedStream.stream(hostToTokenRangeMap)
                .mapKeys(host -> lookUpAddress(nodeMetadataHostMap, host))
                .map(List::stream)
                .map(stream -> stream.collect(Collectors.toSet()))
                .collectToMap();
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:117]: the internal product would add the token range to the map for _one random host_, and then repair all returned token ranges for all hosts.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:101]: I guess this could return a stream to make it nicer. It is probably correct for us to serialise this, especially since we already may have had a bug here?!

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java
Comment (gmaretic) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java:153]: This is probably a leftover from testing?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java
Comment (gmaretic) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java:115]: Also check that the right side is less or equal to (taking into account wrap around)?

**Review by gmaretic — COMMENTED** [review]
Ok, finished the initial review. There is lots of style comments where I basically don't want to just check in the hacky or difficult to parse code from the internal tooling. There is one discrepancy between the javadoc and impl that is inherited from there, and one thing that might be a legit bug, though impl-specific we could be in the clear.

All the building blocks are here, though, so we should be able to get this through in a reasonable amount of time

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:117]: Ah the latter part makes sense, the former part seems weird -- I like the change in any case then, but we should fix the javadoc then

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:120]: sure

File: atlasdb-cassandra/build.gradle
Comment (gsheasby) [reply atlasdb-cassandra/build.gradle:71]: I support this goal :) I think we can get rid of it (at least in atlasdb-cassandra) before merge

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:125]: The protocol version is actually ignored for all types of token, e.g. BlobToken for OPPToken:
```
    @Override
    public ByteBuffer serialize(ByteBuffer value, ProtocolVersion protocolVersion) {
      return value == null ? null : value.duplicate();
    }
```

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java
Comment (gsheasby) [reply atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraRepairEteTest.java:153]: yep, unused -- I'll remove

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:117]: yep, done.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:69]: added a todo

**Review by jeremyk-91 — COMMENTED** [review]
@gmaretic asked me to take a brief look while he looked at transactions3 things. I haven't looked at the tests yet - may not want to action this until after he's had a look

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraServersConfigs.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraServersConfigs.java:159]: Maybe this should just return an `Optional<CqlCapableConfig>` and then you can map it as you wish? This is a little bit strange

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:115]: This is a bit magical, where did we come up with it from?

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraCluster.java
Comment (gmaretic) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraCluster.java:48]: classic IntelliJ

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/LightweightOppToken.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/LightweightOppToken.java:39]: This V3 (and other usages) should live in CssandraConstants

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:58]: I thought IntelliJ displayed this as a comment to me, def annotate 👯

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:50]: This is a constant in CassandraConstants on develop now

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:119]: Annotate

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:153]: This can also be a precondition check instead of a comment

**Review by gmaretic — COMMENTED** [review]
A few more comments, once these are addressed and it's rebased on develop, we should be good

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:66]: added

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlCluster.java:58]: annotating forces you to make it package-private (thanks, ErrorProne!) - I use that when I can, but in several cases it's not possible.

File: atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java
Comment (gsheasby) [reply atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java:63]: yeah, I'm all for reducing side effects. I think the sneaky "peek" option was recommended by @gmaretic in a previous PR.

I'll make the side-effect more explicit, and add a comment (maybe a javadoc) about the exception behaviour.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/LightweightOppToken.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/LightweightOppToken.java:38]: 😞  🇬🇧

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:120]: 🙅 

sadly can't...

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:154]: it already is a precondition check :)

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java:154]: removed the redundant comment

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java:115]: cassandra-driver-core: https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/Token.java#L352

**Review by gmaretic — APPROVED** [review]

**Pull request merged**
